src/axom/multimat/multimat.hpp

```diff
@@ -0,0 +1,765 @@
+                field_mapping == FieldMapping::PER_MAT);
+    const RangeSetType& s =
+      *static_cast<RangeSetType*>(get_mapped_set(field_mapping));
+    Field1D<T>* new_map_ptr = new Field1D<T>(s, T(), stride);
+
+    //copy data
+    int i = 0;
+    for(auto iter = new_map_ptr->begin(); iter != new_map_ptr->end(); iter++)
+      for(auto str = 0; str < stride; ++str) iter(str) = data_arr[i++];
+
+    m_mapVec.push_back(new_map_ptr);
+  }
+
+  m_fieldNameVec.push_back(field_name);
+  m_fieldMappingVec.push_back(field_mapping);
+  m_fieldDataLayoutVec.push_back(data_layout);
+  m_fieldSparsityLayoutVec.push_back(sparsity_layout);
+
+  if(std::is_same<T, int>::value)
+    m_dataTypeVec.push_back(DataTypeSupported::TypeInt);
+  else if(std::is_same<T, double>::value)
+    m_dataTypeVec.push_back(DataTypeSupported::TypeDouble);
+  else if(std::is_same<T, float>::value)
+    m_dataTypeVec.push_back(DataTypeSupported::TypeFloat);
+  else if(std::is_same<T, unsigned char>::value)
+    m_dataTypeVec.push_back(DataTypeSupported::TypeUnsignChar);
+  else
+    m_dataTypeVec.push_back(DataTypeSupported::TypeUnknown);
+
+  SLIC_ASSERT(m_mapVec.size() == m_fieldNameVec.size());
+  SLIC_ASSERT(m_mapVec.size() == m_dataTypeVec.size());
+  SLIC_ASSERT(m_mapVec.size() == m_fieldMappingVec.size());
+  SLIC_ASSERT(m_mapVec.size() == m_fieldDataLayoutVec.size());
+  SLIC_ASSERT(m_mapVec.size() == m_fieldSparsityLayoutVec.size());
+
+  return new_arr_idx;
+}
+
+template <typename T>
+MultiMat::Field1D<T>& MultiMat::get1dField(const std::string& field_name)
+{
+  int fieldIdx = getFieldIdx(field_name);
+
+  if(fieldIdx < 0)
+    throw std::invalid_argument("No field with this name is found");
+
+  if(m_fieldMappingVec[fieldIdx] == FieldMapping::PER_CELL ||
+     m_fieldMappingVec[fieldIdx] == FieldMapping::PER_MAT)
+  {
+    return *dynamic_cast<Field1D<T>*>(m_mapVec[fieldIdx]);
+  }
+  else
+  {
+    SLIC_ASSERT(m_fieldMappingVec[fieldIdx] == FieldMapping::PER_CELL_MAT);
+
+    //Right now we're allowing Field2D (BivariateMap) to be returned as
+    // a Field1D (Map) so it can be accessed like a 1d array, but the
+    // indexing information would be lost.
+    auto* map_2d = dynamic_cast<BivariateMapType<T>*>(m_mapVec[fieldIdx]);
+    return *(map_2d->getMap());
+  }
+}
+
+template <typename T>
+MultiMat::Field2D<T>& MultiMat::get2dField(const std::string& field_name)
+{
+  int fieldIdx = getFieldIdx(field_name);
+
+  if(fieldIdx < 0)
+    throw std::invalid_argument("No field with this name is found");
+
+  SLIC_ASSERT(m_fieldMappingVec[fieldIdx] == FieldMapping::PER_CELL_MAT);
+
+  return *dynamic_cast<Field2D<T>*>(m_mapVec[fieldIdx]);
+}
+
+template <typename T, typename BSetType>
+MultiMat::Field2D<T, BSetType> MultiMat::get2dField(const std::string& field_name)
+{
+  // Get a reference to the unspecialized BMap
+  auto& bmap = get2dField<T>(field_name);
+
+  //create instance of that map
+  int fi = getFieldIdx(field_name);
```

Comment:
Test `fi` and throw exception if less than 0.  (Search for other uses of `getFieldIdx`.)